sled-agent/src/storage_manager.rs

```diff
@@ -664,7 +682,218 @@ impl StorageWorker {
                 "failed to notify nexus, will retry in {:?}", delay;
             );
         };
-        nexus_notifications.push_back(
+        self.nexus_notifications.push_back(
+            backoff::retry_notify(
+                backoff::retry_policy_internal_service_aggressive(),
+                notify_nexus,
+                log_post_failure,
+            )
+            .boxed(),
+        );
+    }
+
+    async fn ensure_using_exactly_these_disks(
+        &mut self,
+        resources: &StorageResources,
+        unparsed_disks: Vec<UnparsedDisk>,
+    ) -> Result<(), Error> {
+        let mut new_disks = HashMap::new();
+
+        // We may encounter errors while parsing any of the disks; keep track of
+        // any errors that occur and return any of them if something goes wrong.
+        //
+        // That being said, we should not prevent access to the other disks if
+        // only one failure occurs.
+        let mut err: Option<Error> = None;
+
+        // Ensure all disks conform to the expected partition layout.
+        for disk in unparsed_disks.into_iter() {
+            match crate::hardware::Disk::new(&self.log, disk).map_err(|err| {
+                warn!(self.log, "Could not ensure partitions: {err}");
+                err
+            }) {
+                Ok(disk) => {
+                    let devfs_path = disk.devfs_path().clone();
+                    new_disks.insert(devfs_path, disk);
+                }
+                Err(e) => {
+                    warn!(self.log, "Cannot parse disk: {e}");
+                    err = Some(e.into());
+                }
+            };
+        }
+
+        let mut disks = resources.disks.lock().await;
```

Comment:
Is this lock automatically released when it goes out of scope?

Reply:
Yup, locks in rust typically implement a `Drop` which releases it

Reply:
That's very cool.